opal/mca/threads/tsd.h

```diff
@@ -0,0 +1,178 @@
+ * Copyright (c) 2008      Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2015-2017 Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+
+#ifndef OPAL_MCA_THREADS_TSD_H
+#define OPAL_MCA_THREADS_TSD_H
+
+#include "opal_config.h"
+
+#include <pthread.h>
+
+#include "opal/constants.h"
+
+BEGIN_C_DECLS
+
+/**
+ * @file
+ *
+ * Thread Specific Datastore Interface
+ *
+ * Functions for providing thread-specific datastore capabilities.
+ */
+
+
+/**
+ * Prototype for callback when tsd data is being destroyed
+ */
+typedef void (*opal_tsd_destructor_t)(void *value);
+
+#if defined(DOXYGEN)
+
+/**
+ * Typedef for thread-specific data key
+ */
+typedef void* opal_tsd_key_t;
+
+
+/**
+ * Delete a thread-specific data key
+ *
+ * Delete a thread-specific data key previously returned by
+ * opal_tsd_key_create().  The destructor associated with the key is
+ * not fired in any thread and memory cleanup is the responsibility of
+ * the caller.
+ *
+ * @note Unlike pthread_key_delete, this function should not be called
+ * from within a destructor.  It can not be universally supported at
+ * this time.
+ *
+ * @param key[in]       The key for accessing thread-specific data
+ *
+ * @retval OPAL_SUCCESS Success
+ * @retval EINVAL       Invalid key
+ */
+OPAL_DECLSPEC int opal_tsd_key_delete(opal_tsd_key_t key);
+
+
+/**
+ * Set a thread-specific data value
+ *
+ * Associates value with key in the current thread.  The value for the
+ * key in other threads is not changed.  Different threads may assign
+ * different values to the same key.
+ *
+ * @note This function should not be called within
+ * opal_tsd_key_delete().
+ *
+ * @param key[in]       Thread specific data key to modify
+ * @param value[in]     Value to associate with key
+ *
+ * @retval OPAL_SUCCESS Success
+ * @retval ENOMEM       Insufficient memory exists to associate the
+ *                      value with the key
+ * @retval EINVAL       Invalid key
+ */
+OPAL_DECLSPEC int opal_tsd_setspecific(opal_tsd_key_t key, void *value);
+
+
+/**
+ * Get a thread-specific data value
+ *
+ * Get the data associated with the given key, as set by
+ * opal_tsd_setspecific().  If opal_tsd_setspecific() hasn't been
+ * called in the current thread with the given key, NULL is returned
+ * in valuep.
+ *
+ * @param key[in]       Thread specific data key to modify
+ * @param value[out]     Value to associate with key
+ *
+ * @retval OPAL_SUCCESS Success
+ * @retval ENOMEM       Insufficient memory exists to associate the
+ *                      value with the key
+ * @retval EINVAL       Invalid key
+ */
+OPAL_DECLSPEC int opal_tsd_getspecific(opal_tsd_key_t key, void **valuep);
+
+#else
+
+typedef pthread_key_t opal_tsd_key_t;
+
+static inline int
+opal_tsd_key_delete(opal_tsd_key_t key)
+{
+    return pthread_key_delete(key);
```

Comment:
These inline functions directly calling pthread functions seem to be an abstraction violation.